{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Series\n",
    "\n",
    "Given a string of digits, output all the contiguous substrings of length `n` in\n",
    "that string.\n",
    "\n",
    "For example, the string \"49142\" has the following 3-digit series:\n",
    "\n",
    "- 491\n",
    "- 914\n",
    "- 142\n",
    "\n",
    "And the following 4-digit series:\n",
    "\n",
    "- 4914\n",
    "- 9142\n",
    "\n",
    "And if you ask for a 6-digit series from a 5-digit string, you deserve\n",
    "whatever you get.\n",
    "\n",
    "Note that these series are only required to occupy *adjacent positions*\n",
    "in the input; the digits need not be *numerically consecutive*.\n",
    "\n",
    "## Exception messages\n",
    "\n",
    "Sometimes it is necessary to raise an exception. When you do this, you should include a meaningful error message to\n",
    "indicate what the source of the error is. This makes your code more readable and helps significantly with debugging. Not\n",
    "every exercise will require you to raise an exception, but for those that do, the tests will only pass if you include\n",
    "a message.\n",
    "\n",
    "To raise a message with an exception, just write it as an argument to the exception type. For example, instead of\n",
    "`raise Exception`, you shold write:\n",
    "\n",
    "```python\n",
    "raise Exception(\"Meaningful message indicating the source of the error\")\n",
    "```\n",
    "\n",
    "## Submitting Exercises\n",
    "\n",
    "Note that, when trying to submit an exercise, make sure the solution is in the `$EXERCISM_WORKSPACE/python/series` directory.\n",
    "\n",
    "You can find your Exercism workspace by running `exercism debug` and looking for the line that starts with `Workspace`.\n",
    "\n",
    "For more detailed information about running tests, code style and linting,\n",
    "please see the [help page](http://exercism.io/languages/python).\n",
    "\n",
    "## Source\n",
    "\n",
    "A subset of the Problem 8 at Project Euler [http://projecteuler.net/problem=8](http://projecteuler.net/problem=8)\n",
    "\n",
    "## Submitting Incomplete Solutions\n",
    "It's possible to submit an incomplete solution so you can see how others have completed the exercise.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def slices(series, length):\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Tests for the series exercise\n",
    "\n",
    "Implementation note:\n",
    "The slices function should raise a ValueError with a meaningful error\n",
    "message if its length argument doesn't fit the series.\n",
    "\"\"\"\n",
    "import unittest\n",
    "\n",
    "\n",
    "\n",
    "class SeriesTest(unittest.TestCase):\n",
    "    def test_slices_of_one(self):\n",
    "        self.assertEqual(\n",
    "            slices(\"01234\", 1),\n",
    "            [[0], [1], [2], [3], [4]], )\n",
    "\n",
    "    def test_slices_of_two(self):\n",
    "        self.assertEqual(\n",
    "            slices(\"97867564\", 2),\n",
    "            [[9, 7], [7, 8], [8, 6], [6, 7], [7, 5], [5, 6], [6, 4]], )\n",
    "\n",
    "    def test_slices_of_three(self):\n",
    "        self.assertEqual(\n",
    "            slices(\"97867564\", 3),\n",
    "            [[9, 7, 8], [7, 8, 6], [8, 6, 7], [6, 7, 5], [7, 5, 6], [5, 6, 4]],\n",
    "        )\n",
    "\n",
    "    def test_slices_of_four(self):\n",
    "        self.assertEqual(\n",
    "            slices(\"01234\", 4),\n",
    "            [[0, 1, 2, 3], [1, 2, 3, 4]], )\n",
    "\n",
    "    def test_slices_of_five(self):\n",
    "        self.assertEqual(\n",
    "            slices(\"01234\", 5),\n",
    "            [[0, 1, 2, 3, 4]], )\n",
    "\n",
    "    def test_overly_long_slice(self):\n",
    "        with self.assertRaisesWithMessage(ValueError):\n",
    "            slices(\"012\", 4)\n",
    "\n",
    "    def test_overly_short_slice(self):\n",
    "        with self.assertRaisesWithMessage(ValueError):\n",
    "            slices(\"01234\", 0)\n",
    "\n",
    "    # Utility functions\n",
    "    def setUp(self):\n",
    "        try:\n",
    "            self.assertRaisesRegex\n",
    "        except AttributeError:\n",
    "            self.assertRaisesRegex = self.assertRaisesRegexp\n",
    "\n",
    "    def assertRaisesWithMessage(self, exception):\n",
    "        return self.assertRaisesRegex(exception, r\".+\")\n",
    "\n",
    "\n",
    "unittest.main(argv=[''], exit=False)\n"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 2
}
